validphys2/src/validphys/pseudodata.py

```diff
@@ -148,77 +148,29 @@ def make_replica(groups_dataset_inputs_loaded_cd_with_cuts, replica_mcseed, genr
     name_salt = "-".join(i.setname for i in groups_dataset_inputs_loaded_cd_with_cuts)
     name_seed = int(hashlib.sha256(name_salt.encode()).hexdigest(), 16) % 10 ** 8
     rng = np.random.default_rng(seed=replica_mcseed+name_seed)
+    #construct covmat
+    covmat = dataset_inputs_total_covmat
+    covmat_sqrt = sqrt_covmat(covmat)
 
     # The inner while True loop is for ensuring a positive definite
     # pseudodata replica
     while True:
         pseudodatas = []
-        special_add = []
-        special_mult = []
-        mult_shifts = []
         check_positive_masks = []
         for cd in groups_dataset_inputs_loaded_cd_with_cuts:
```

Comment:
The changes I did for the additive errors might be savaged to remove this loop: https://github.com/NNPDF/nnpdf/pull/1544/files

which will speed up even more the generation of replicas (which is important given the amount of negative pseudodata we get)

Reply:
I don't understand, we don't use anymore the additive and multiplicative errors here. We only loop to collect the pseudodata, right?

Reply:
Yes, but you are still looping over all datasets for every iteration of the outer `while`. My point is that you now can do that just once as I did here https://github.com/NNPDF/nnpdf/pull/1544/files and extract all the central values and the positivity mask.